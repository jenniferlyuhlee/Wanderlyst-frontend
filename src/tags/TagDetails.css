.TagDetails{
    margin-bottom: 3rem;
}

.TagName{
    background-color: rgba(0, 0, 0, 0.2);
    padding: 5%;
    width: 100%;
    height: 10%;
}

.TagName-bg{
    background-size: cover;
    background-repeat: no-repeat;
    color: white;
}

/* Dynamic classes for each tag banner*/

/* Image sourced from:
https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcStEulQnBJQLxrIsRTaYdZ94eR2XjjF4Vs0Xg&s */
.TagName-bg-Foodie{
    background-image: url("../assets/tag-banners/foodie.jpg");
    background-position: 0% 60%;
}
/* Image sourced from:
https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcRSujr7zcu9Ql5pQKfqq0EWrkiGrgrJCMAGPQ&s */
.TagName-bg-Explorer{
    background-image: url("../assets/tag-banners/explorer.jpg");
    background-position: 0% 55%;
}
/* Image sourced from:
https://media.timeout.com/images/105266065/image.jpg */
.TagName-bg-Shopaholic{
    background-image: url("../assets/tag-banners/shopaholic.jpg");
    background-position: center;
}
/* Image sourced from: 
https://www.travelandleisure.com/thmb/QK4CuWpFdm2dR3NDznbBQtsAoN0=/1500x0/filters:no_upscale():max_bytes(150000):strip_icc()/northern-lights-kirkjufell-mountain-snaefellsnes-iceland-ICELANDLIGHTS1218-824f48715748425f828f05aa2a28dfe0.jpg */
.TagName-bg-Naturalist{
    background-image: url("../assets/tag-banners/naturalist.jpg");
    background-position: 0% 40%;
}
/* Image sourced from:
https://media.cntraveler.com/photos/5a84852c7211e1293261ceb8/16:9/w_2560,c_limit/Skydive-Dubai_2018_GettyImages-643320488.jpg */
.TagName-bg-Daredevil{
    background-image: url("../assets/tag-banners/daredevil.jpeg");
    background-position: 0% 45%;
}
/* Image sourced from:
https://www.intrepidtravel.com/v3/assets/blt0de87ff52d9c34a8/blt8cf727b4373c6aaa/6322fd2c02b4e92fd30d0030/Intrepid_Travel-Japan-Miyajima_1!.jpg?branch=prd */
.TagName-bg-Flâneur{
    background-image: url("../assets/tag-banners/flâneur.jpg");
    background-position: 0% 80%;
}
/* Image sourced from:
https://media.worldnomads.com/travel-safety/thailand/nightlife-pattaya-social.jpg */
.TagName-bg-Reveller{
    background-image: url("../assets/tag-banners/reveller.jpg");
    background-position: 0% 65%;
}
/* Image sourced from: 
https://live.staticflickr.com/2102/2035748576_d3735ec49c_h.jpg */
.TagName-bg-Aesthete{
    background-image: url("../assets/tag-banners/aesthete.jpg");
    background-position: 0% 15%;
}
/* Image sourced from:
https://i.natgeofe.com/n/fba4cbb4-6a7f-4a4d-9e32-9f691cef8856/higts_southafrica_8502902-Edit_HR_square.jpg */
.TagName-bg-Relaxer{
    background-image: url("../assets/tag-banners/relaxer.jpg");
    background-position: 0% 60%;
}
/* Image sourced from:
https://shuttermuse.com/wp-content/uploads/2020/04/african-wildlife-safari-elephant-1400px.jpg */
.TagName-bg-Photographer{
    background-image: url("../assets/tag-banners/photographer.jpg");
    background-position: 0% 55%;
}

