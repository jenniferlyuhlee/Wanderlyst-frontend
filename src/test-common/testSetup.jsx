import React, { useState } from "react";
import UserContext from "../auth-user/UserContext";
import WanderlystApi from "../utils/api";
import '@testing-library/jest-dom';

const demoUser = {
  username: "testadmin", 
  firstName: "testfirst",
  lastName: "testlast",
  location: "testloc",
  bio: "testbio",
  profilePic: null,
  isAdmin: true,
};

const demoTags = [
  {id: 1, name: 'Foodie', description: 'Savor the world one bite at a time. Dive into a journey packed with delicious local cuisines, hidden culinary gems, and must-try dishes that will leave your taste buds tingling.'},
  {id: 2, name: 'Explorer', description: 'Step off the beaten path and uncover the unknown. From ancient ruins to vibrant cityscapes, let your curiosity guide you through unforgettable experiences.'}
]

const UserProvider = ({ children, 
                        currUser = demoUser, 
                        setCurrUser, 
                        tags = demoTags,
                        logout,
                        hasLikedItin,
                        toggleLike }) => {
  const [user, setUser] = useState(currUser);

  return (
    <UserContext.Provider value={{ currUser: user, 
                                  setCurrUser: setUser, 
                                  tags, logout, 
                                  hasLikedItin, toggleLike}}>
      {children}
    </UserContext.Provider>
  );
};

export { UserProvider };