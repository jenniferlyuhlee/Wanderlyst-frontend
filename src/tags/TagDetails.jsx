import React, { useState, useEffect } from "react";
import { useParams } from "react-router-dom";
import WanderlystApi from "../utils/api";
import ItinCard from "../itineraries/ItinCard";
import Loading from "../shared/Loading";
import "./TagDetails.css"

/** Tag Details Page Component
 * Displays tag details and associated itineraries
 * Params: name
 */

function TagDetails(){
    const { name } = useParams();
    // state set to null to use loading spinner
    const [tag, setTag] = useState(null);
    // error state for invalid param tag name
    const [error, setError] = useState(false);

    // fetches tag details from API upon initial load
    useEffect(function getTagDetails() {
        async function getTag(){
            try{
                setTag(await WanderlystApi.getTag(name));
            }
            catch(err){
                setError(true);
            }
        }
        getTag();
    }, [name]);

    if(error) return <p className="NotFound"><i>Sorry this tag doesn't exist.</i></p>

    if(!tag) return <Loading />

    return(
        <div className = "TagDetails">
            <div className = {`TagName-bg TagName-bg-${tag.name}`}>
                <h1 className = "text-center display-4 wanderlyst fw-medium TagName">
                    {tag.name}
                </h1>
            </div>
            <div className="container">
                <p className="m-4">{tag.description}</p>
                {tag.itineraries.length > 0 ?
                <div>
                    {tag.itineraries.map(i => (
                        <ItinCard key = {i.id} itinerary={i}/>
                    ))}
                </div>
                :
                <p className="text-center fst-italic">
                    <i>Sorry, no itineraries have been tagged yet.</i>
                </p>
                }
            </div>
        </div>
    )
}

export default TagDetails;