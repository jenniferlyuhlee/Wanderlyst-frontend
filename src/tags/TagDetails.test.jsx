import React from "react";
import { render, fireEvent, waitFor } from "@testing-library/react";
import { UserProvider } from "../test-common/testSetup";
import { MemoryRouter} from "react-router-dom";
import WanderlystApi from "../utils/api";
import TagDetails from "./TagDetails";
import { beforeEach, afterEach, expect, describe, test } from "vitest";

describe("TagDetails component", () => {
    // mock api method to get testTag
    beforeEach(() => {
        vi.spyOn(WanderlystApi, "getTag").mockResolvedValue({
            name: "testTag",
            description: "Tag for testing.",
            itineraries: []
        })
    });
      
    // Restore mocks after tests
    afterEach(() => {
        vi.restoreAllMocks();
    });

    test("renders without crashing", async () => {
        await waitFor(() => { render(
            <MemoryRouter initialEntries={['/tags/testTag']}>
                <UserProvider>
                    <TagDetails/>
                </UserProvider>
            </MemoryRouter>)
        });
    });

    test("matches snapshot", async () => {
        const {getByText, asFragment} = render(
            <MemoryRouter initialEntries={['/tags/testTag']}>
                <UserProvider>
                    <TagDetails/>
                </UserProvider>
            </MemoryRouter>)

        // wait and check that tag is displayed
        await waitFor(() => {
            expect(getByText("Tag for testing.")).toBeInTheDocument();
        });
        expect(asFragment()).toMatchSnapshot();
       
    });
    test("displays loading spinner initially", () => {
        const { container } = render(
            <MemoryRouter initialEntries={['/tags/testTag']}>
                <UserProvider>
                    <TagDetails />
                </UserProvider>
            </MemoryRouter>
        );

        // Expect loading spinner to appear before the API data is loaded
        expect(container.querySelector(".fa-spinner")).toBeInTheDocument();
    });

    test("displays error message if tag does not exist", async () => {
        // Mock API to throw an error
        WanderlystApi.getTag.mockRejectedValueOnce(new Error("Tag not found"));

        const {getByText} = render(
            <MemoryRouter initialEntries={['/tags/nonexistentTag']}>
                <UserProvider>
                    <TagDetails />
                </UserProvider>
            </MemoryRouter>
        );

        // Wait and check for the error message to appear
        await waitFor(() => {
            expect(getByText("Sorry this tag doesn't exist.")).toBeInTheDocument();
        });
    });

});