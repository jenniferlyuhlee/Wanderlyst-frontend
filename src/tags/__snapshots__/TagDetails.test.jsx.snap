// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TagDetails component > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="TagDetails"
  >
    <div
      class="TagName-bg TagName-bg-testTag"
    >
      <h1
        class="text-center display-4 wanderlyst fw-medium TagName"
      >
        testTag
      </h1>
    </div>
    <div
      class="container"
    >
      <p
        class="m-4"
      >
        Tag for testing.
      </p>
      <p
        class="text-center fst-italic"
      >
        <i>
          Sorry, no itineraries have been tagged yet.
        </i>
      </p>
    </div>
  </div>
</DocumentFragment>
`;
